### ISSUE TITLE
AsyncSSH and IOS-XR 7.5.4 not clearing interfaces counters, clear confirmation not sent

### ISSUE THREAD

**Issue by alfanaja** [ronf/asyncssh#646]
Hello AsyncSSH team  we are doing testings, we found when using Async 2.8.1 and IOS-XR 7.5.4  our  script must clear interface counters  with the command "clear counters interface x/x/x/x".  we instruct the script to go and delete counters for all considerable amount of  interfaces. 

sometimes the interface counters don't in fact clear, and we were able to reproduce using their script. Our debugging seems to suggest that there is an issue with regards to the confirm output. Asyncssh waits until is sees this prompt, and then sends a stdin.write('\r\r') and then disconnects. 

When we disabled the confirmation in the router config, we don't see the problem.
 
Are you aware of some similar behavior?. 

we can arrange  access to the lab environment to test with the IOS-XR router. 

AsyncSSH Version:

[anonymous@anonymous-devbox:~/netinfra/x]$ pip3 freeze | grep asyncssh
asyncssh==2.8.1

**Comment by smilstea** [comment]
For reference here is the code we are using after we detect the confirm prompt:
            stdin.write("\r\r")
            stdin.write_eof()
            await stdin.drain() 

Is this the correct way to send a confirm? Our theory is that the session is disconnecting immediately after sending the carriage return but has a racecondition causing the command to not be read by the parser in ios xr correctly.

**Comment by ronf** [comment]
I would recommend against doing a write_eof() here, at least not until after you have finished reading all of the output that you want to read from the device. Even though EOF in SSH does allow "half closed" connections, many SSH implementations treat receiving EOF as a full close, where the connection is closed before all of the output of the last command is sent.

Ideally, if you are scanning the output for prompts, you should really wait for the prompt both at the beginning (before sending a command), and then again after sending the command to detect the end of the output. Only after getting the full output and another prompt would you want to do the write_eof(), or perhaps just do a close() and await on wait_closed() at that point.

I'm also not quite sure why you'd be writing two carriage returns here. Are you doing this to try to get an additional prompt output?

**Comment by smilstea** [comment]
Thanks for the explanation.

**Comment by ronf** [comment]
Closing due to inactivity. Feel free to reopen this or open a new issue if you need anything else.
